doc/guides/kconfig/tips.rst

```diff
@@ -51,6 +46,48 @@ without prompts in mind.
 See the `optional prompts`_ section for a way to deal with settings that are
 fixed on some machines and configurable on other machines.
 
+What not to turn into Kconfig options
+*************************************
+
+In Zephyr, Kconfig configuration is done after selecting a target board. In
+general, it does not make sense to use Kconfig for a value that corresponds to
+a fixed machine-specific setting. Usually, such settings should be handled via
+:ref:`devicetree <dt-guide>` instead.
+
+In particular, avoid adding new Kconfig options of the following types:
+
+Options enabling individual devices
+===================================
+
+Existing examples like :option:`CONFIG_SPI_0` and :option:`CONFIG_I2C_1` were
+introduced before Zephyr supported devicetree, and new cases are discouraged.
+See :ref:`dt-create-devices` for details on how to do this with devicetree
+instead.
+
+Options that specify a device in the system by name
+===================================================
+
+For example, if you are writing an I2C device driver, avoid creating an option
+named ``MY_DEVICE_I2C_BUS_NAME`` for specifying the bus node your device is
+controlled by. See :ref:`dt-drivers-that-depend` for alternatives.
+
+Similarly, if your application depends on a hardware-specific PWM device to
+control an RGB LED, avoid creating an option like ``MY_PWM_DEVICE_NAME``. See
+:ref:`dt-apps-that-depend` for alternatives.
+
+Options that specify fixed hardware configuration
+=================================================
+
+For example, avoid Kconfig options specifying a GPIO pin.
+
+An alternative applicable to device drivers is to define a GPIO specifier with
+type phandle-array in the device binding, and using the
+:ref:`devicetree-gpio-api` devicetree API from C. Similar advice applies to
+other cases where devicetree.h provides :ref:`devicetree-hw-api` for referring
+to other nodes in the system. Search the source code for drivers using these
+APIs for examples.
```

Comment:
Might be worth pointing to `tests/drivers/gpio/gpio_basic_api` which provides an application-specific binding to identify board-specific properties, which is a step beyond using aliases.